### PR TITLE
GameServer: Create Local Spring Profile for Local Testing

### PR DESCRIPTION
**Pull request by klokailo**
## Background
- Previously, the gameserver application would fail to run unless it was deployed within our K8s cluster with agones installed as well.
  - This is for two reasons:
    - The `AgonesHook` will fail to connect to the localhost Kubernetes API when trying to retrieve gameserver information from the Agones Controller Service.
      - It will succeed in setting up the Agones Hook webserver though, but this is unnecessary for outside-of-cluster deployment.
    - The `RedisStorageService` will not be able to connect to a redis database.
## Goal
- We want to have a different spring boot profile called `local`, which when set, allows our gameserver to start listening on a local port for socket connections.
- We should consider what these services are used for:
  - `AgonesHook` specifically communicates with Agones in the cluster, reporting health status and also gets the the Cluster ExternalIP and dynamic agones-assigned port for this container.
    - We do not need this for local testing, since we do not need dynamic port allocation if we are running just a single gameserver backend.
  - `RedisStorageService` stores things like client sessions and gameserver info, but crucially, is <i>shared</i> across all gameservers in the cluster.
    - This storage remains necessary for local deployment, but the need to host it in an external shared database no longer exists since there would only be one gameserver accessing it.
## This PR
- For the Redis storage, we know it is not useful when we deploy locally. Therefore, we can simply just use an in-memory hashmap solution instead.
  - We create two implementations of the same GameServerStorage interface, then create a bean with the type corresponding to our environment.
- Creates the `local` springboot profile, which:
  - Sets a static port routing for our service on port 7654 (default port for gameservers)
  - Disables the hook into agones, and instead replaces the information on `currentGameServer` with `localhost:7654`
  - Sets our storage implementation to use local storage instead of redis.
## Running Locally
- In intelliJ, Run -> Edit Run Configurations -> SpringBoot -> GameServer -> Spring Active Profiles -> add "local"
- Upon executing this run configuration, the game server will listen locally on port `7654`.

### CODE REVIEW

File: gameserver/src/main/java/dev/totallyspies/spydle/gameserver/storage/CurrentGameServerConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+                .roomId("12345") // TODO
+                .publicRoom(false) // TODO
+                .state(GameServer.State.WAITING)
+                .build();
+    }
+
+    @EventListener(ApplicationReadyEvent.class)
+    public void onApplicationReady() {
+        GameServer currentGameServer = applicationContext.getBean(GameServer.class);
+        storage.storeGameServer(currentGameServer);
```
Comment (chopikus) [inline gameserver/src/main/java/dev/totallyspies/spydle/gameserver/storage/CurrentGameServerConfiguration.java:58]: does it have to be synchronized?

**Comment by chopikus** [comment]
```For the Redis storage, we know it is not useful when we deploy locally. Therefore, we can simply just use an in-memory hashmap solution instead. ```

I don't understand how Redis can be useful in one profile but not useful in another. What if game logic wants to store something?

The great advantage of Redis over ConcurrentHashMap is that we don't have to worry over synchronization of values, especially if we the alternative is to have something complicated like ConcurrentHashMap<String, List<String>>.

If we run a developer profile, we still have a hashmap that can be accessed by multiple threads (f.e. running 2 requests at the same time), and even using a ConcurrentHashMap can be tricky.

Take a look at this article: https://dzone.com/articles/concurrenthashmap-isnt-always-enough

**Comment by chopikus** [comment]
I don't think keeping two different implementations of storing the same thing is a good idea. Is there any issue with connecting to the local redis server installed on a developer machine?

**Comment by klokailo** [comment]
@chopikus Currently, redis is being used to store information that is shared between multiple gameservers (and maybe matchmaker). Things like which gameserver has which roomId, which client is connected to which gameserver, etc.
In local deployment, we don't have anything else to communicate with, so we no longer need to be external.
But to answer your second question, no, there isn't a problem with connecting to a redis database for local deployment. You would set in the `application.properties` the `storage.type=redis`, and `redis.host` + `redis.port` accordingly.

But I think this misundetstanding is because you are thinking about using redis for stuff that is different from what I have described. Which is fine, of course it can be used for many things. 
Although I think that the concern of performance of the lock-on-read that ConcurrentHashMap performs is likely not significant for our game.
What are you thinking of storing in the database?

**Comment by chopikus** [comment]
I'm thinking to use it for storing data for the game logic, like list of words to generate from, a substring for the current turn, etc.

These things can be also accessed by multiple threads I think?

On the ConcurrentHashmap, it is not a concern on performance (there is no problem there), but a concern on data integrity. 

The article shows a simple mistake that could be made that could mess up a value for a specific key in a map.

**Comment by klokailo** [comment]
@chopikus I think that using an external database in this case would mildly over-engineering it.
Just put the logic that you use to write to the storage map in a `synchronized` block ([java keyword](https://stackoverflow.com/questions/1085709/what-does-synchronized-mean)). 

Also I think the `GameServerStorage` interface should not include <b>stateful</b> properties of our game: things like the current word to guess and so on. A separate datastore is the best solution. That separate datastore could be whatever you want, so if you don't like `synchronized`, you can use a `ConcurrentHashMap` with `computeIfAbsent`.

**Comment by klokailo** [comment]
@chopikus Any other comments or ideas?

**Comment by chopikus** [comment]
I don't understand why we have Redis for the matchmaker in this case? Why are we using 2 different solutions when we can use just 1 everywhere

```
Also I think the GameServerStorage interface should not include stateful properties of our game: things like the current word to guess and so on. A separate datastore is the best solution. 
```

Why not?

**Comment by klokailo** [comment]
> I don't understand why we have Redis for the matchmaker in this case? Why are we using 2 different solutions when we can use just 1 everywhere

I'm not sure I understand this question. Both the GameServers and the Matchmakers are all connected to a single redis database, where they can store information that is pertinent to all of them.

What I am saying is that the `GameServerStorage` interface, although you could rename it, currently represents this shared storage with other gameservers and matchmakers.

The stateful data that the redis database currently stores is data that is relevant to the system as a whole. If we have data that we need to store that is only relevant to a single gameserver, then it should be stored on that gameserver instance. (I'm sure this follows from the POLP or something. Definitely standard design choice.)

I am not sure why this would be an issue: For the game logic data, using a `ConcurrentHashMap` with `synchronized` blocks when necessary is:
- Not really a "second solution" when it comes to storing data, since it is different data with a different lifecycle and different audicence that you are storing, so it makes sense to store it separately
- Not compromising data integrity if programmed correctly
- Definitely not an issue of performance as I doubt the hashmap would be called that frequently.

If this still doesn't make sense then we can meet tomorrow and discuss it. But I think it is important we proceed with actually writing game logic and not get too stuck up on smaller details.

(Small aside: the only real benefit I see to storing game logic data in the shared datastore is that it makes our gameservers, truly, 100%, stateless. But the only way we would benefit from this is if I wrote some logic that is able to transfer players from one gameserver to another, if the original one failed or something like that. But this approach goes against all the principals of Agones: clients connect to a single gameserver, game servers are dynamically allocated, gameservers are not arbitrary in network routing, etc. These principals exist for good reason: because it is good design when it comes to making games. GameServers should not be completely stateless with respect to their own logic. For performant games, you can easily see how breaking these principals would be a problem. Try implementing a first-person-shooter using stateless logic.)

**Review by chopikus — APPROVED** [review]

**Pull request merged**